File: BaseSDK/Ubuntu16.04/Samples/Base/DS86/MultiConnection/MultiConnection.cpp
#include <thread>
#include <iostream>
#include "Scepter_api.h"

#define frameSpace 10
using namespace std;

int main()
{
	cout << "---MultiConnection---" << endl;

	uint32_t deviceCount = 0;
	ScStatus status = SC_OTHERS;
	ScDeviceHandle *deviceHandle = nullptr;
	ScDeviceInfo* pDeviceListInfo = nullptr;
	ScFrameReady frameReady = { 0 };
	ScFrame depthFrame = { 0 };

	status = scInitialize();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scInitialize] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scInitialize] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	do
	{
		status = scGetDeviceCount(&deviceCount, 3000);
		if (status == ScStatus::SC_OK)
		{
			cout << "[scGetDeviceCount] success, ScStatus(" << status << "). The device count is " << deviceCount << endl;
		}
		else
		{
			cout << "[scGetDeviceCount] fail, ScStatus(" << status << ")." << endl;
			return -1;
		}
	} while (deviceCount < 2);

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceInfoList] success, ScStatus(" << status << ").";
	}
	else
	{
		cout << "[scGetDeviceInfoList] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return -1;
	}

	cout << endl;
	for (int i = 0; i < deviceCount; i++)
	{
		cout << " The device index: " << i << ", <serialNumber>: " << pDeviceListInfo[i].serialNumber
			<< ", <ip>: " << pDeviceListInfo[i].ip << ", <status>: " << pDeviceListInfo[i].status << endl;
	}

	deviceHandle = new ScDeviceHandle[deviceCount];
	for (int i = 0; i < deviceCount; i++)
	{
		status = scOpenDeviceBySN(pDeviceListInfo[i].serialNumber, &deviceHandle[i]);
		if (status == ScStatus::SC_OK)
		{
			cout << "[scOpenDeviceBySN] success, ScStatus(" << status << "). The device index: " << i << ", <serialNumber>: " << pDeviceListInfo[i].serialNumber << endl;
		}
		else
		{
			cout << "[scOpenDeviceBySN] fail, ScStatus(" << status << "). The device index: " << i << ", <serialNumber>: " << pDeviceListInfo[i].serialNumber << endl;
			delete[] pDeviceListInfo;
			pDeviceListInfo = NULL;
			delete[] deviceHandle;
			deviceHandle = NULL;
			return -1;
		}
	}
	delete[] pDeviceListInfo;
	pDeviceListInfo = NULL;

	for (int i = 0; i < deviceCount; i++)
	{
		status = scStartStream(deviceHandle[i]);
		if (status == ScStatus::SC_OK)
		{
			cout << "[scStartStream] success, ScStatus(" << status << "). The device index: " << i << endl;
		}
		else
		{
			cout << "[scStartStream] fail, ScStatus(" << status << "). The device index: " << i << endl;
			delete[] deviceHandle;
			deviceHandle = NULL;
			return -1;
		}
	}

	//Wait for the device to upload image data.
	this_thread::sleep_for(chrono::milliseconds(1000));

	for (int j = 0; j < frameSpace; j++)
	{
		for (int i = 0; i < deviceCount; i++)
		{
			if (deviceHandle[i])
			{
				status = scGetFrameReady(deviceHandle[i], 1200, &frameReady);
				if (status == ScStatus::SC_OK)
				{
				}
				else
				{
					cout << "[scGetFrameReady] fail, ScStatus(" << status << "). The device index: " << i << endl;
					continue;
				}

				if (1 == frameReady.depth)
				{
					status = scGetFrame(deviceHandle[i], SC_DEPTH_FRAME, &depthFrame);
					if (status == ScStatus::SC_OK)
					{
						cout << "[scGetFrame] success, ScStatus(" << status << "). The device index: " << i << ", SC_DEPTH_FRAME <frameIndex>: " << depthFrame.frameIndex << endl;
					}
					else
					{
						cout << "[scGetFrame] fail, ScStatus(" << status << "). The device index: " << i << endl;
					}
				}
			}
		}
	}

	for (int i = 0; i < deviceCount; i++)
	{
		status = scStopStream(deviceHandle[i]);
		if (status == ScStatus::SC_OK)
		{
			cout << "[scStopStream] success, ScStatus(" << status << "). The device index: " << i << endl;
		}
		else
		{
			cout << "[scStopStream] fail, ScStatus(" << status << "). The device index: " << i << endl;
			delete[] deviceHandle;
			deviceHandle = NULL;
			return -1;
		}

		status = scCloseDevice(&deviceHandle[i]);
		if (status == ScStatus::SC_OK)
		{
			cout << "[scCloseDevice] success, ScStatus(" << status << "). The device index: " << i << endl;
		}
		else
		{
			cout << "[scCloseDevice] fail, ScStatus(" << status << "). The device index: " << i << endl;
			delete[] deviceHandle;
			deviceHandle = NULL;
			return -1;
		}
	}
	delete[] deviceHandle;
	deviceHandle = NULL;

	status = scShutdown();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scShutdown] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scShutdown] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}
	cout << "---End---" << endl;

	return 0;
}

File: MultilanguageSDK/Python/Samples/NYX650/MultiConnection/MultiConnection.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time
print("---MultiConnection---")
camera = ScepterTofCam()

camera_count = 0
while camera_count < 2:
    camera_count = camera.scGetDeviceCount(3000)
    print("[scGetDeviceCount] success, ScStatus(0), The device count is {}".format(camera_count))

cameras = []

ret, device_infolist=camera.scGetDeviceInfoList(camera_count)
if ret==0:
    print("[scGetDeviceInfoList] success, ScStatus({})".format(ret))
    for i in range(camera_count): 
        print(" The device index {}, <serialNumber>: {} , <ip>: {} , <status>: {}".format(str(i), str(device_infolist[i].serialNumber.decode()), str(device_infolist[i].ip.decode()), str(device_infolist[i].status)))
        cam = ScepterTofCam()
        ret = cam.scOpenDeviceBySN(device_infolist[i].serialNumber)
        if  ret == 0:
            print('[scOpenDeviceBySN] success ScStatus({}). The device index {} , <serialNumber>: {}'.format(str(ret), str(i), str(device_infolist[i].serialNumber.decode())))
            cameras.append(cam)
        else:
            print('[scOpenDeviceBySN] fail ScStatus({}). The device index {} , <serialNumber>: {}'.format(str(ret), str(i), str(device_infolist[i].serialNumber.decode())))
            exit(1)
else:
    print("[scGetDeviceInfoList] fail, ScStatus({})".format(ret))
    exit(1)  

for i in range(camera_count): 
    ret = cameras[i].scStartStream()       
    if  ret == 0:
        print('[scStartStream] success ScStatus({}).  The device index {}'.format(str(ret), str(i)))
    else:
        print('[scStartStream] fail ScStatus({}).  The device index {}'.format(str(ret), str(i)))
        exit(1)

#Wait for the device to upload image data.
time.sleep(1)

# show image 

for ind in range(10):
    for i in range(camera_count): 
        ret, frameready = cameras[i].scGetFrameReady(c_uint16(1200))
        if  ret != 0:
            print('[scGetFrameReady] fail ScStatus({}).'.format(str(ret)))
            continue
        if  frameready.depth:      
            ret,depthframe = cameras[i].scGetFrame(ScFrameType.SC_DEPTH_FRAME)
            if  ret == 0:
                print('[scGetFrame] success ScStatus({}). The device index: {}, SC_DEPTH_FRAME <frameIndex>: {}'.format(str(ret), str(i), str(depthframe.frameIndex)))
            else:
                print('[scGetFrame] fail ScStatus({}). The device index: {}, SC_DEPTH_FRAME <frameIndex>: {}'.format(str(ret), str(i), str(depthframe.frameIndex)))
        if  frameready.ir:
            ret,irframe = cameras[i].scGetFrame(ScFrameType.SC_IR_FRAME)
            if  ret == 0:
                print('[scGetFrame] success ScStatus({}). The device index: {}, SC_IR_FRAME <frameIndex>: {}'.format(str(ret), str(i), str(irframe.frameIndex)))
            else:
                print('[scGetFrame] fail ScStatus({}). The device index: {}, SC_IR_FRAME <frameIndex>: {}'.format(str(ret), str(i), str(irframe.frameIndex)))

for i in range(camera_count): 
    
    ret = cameras[i].scStopStream()       
    if  ret == 0:
        print('[scStopStream] success ScStatus({}). The device index: {}'.format(str(ret), str(i)))
    else:
        print('[scStopStream] fail ScStatus({}). The device index: {}'.format(str(ret), str(i)))
        exit(1)

    ret = cameras[i].scCloseDevice()       
    if  ret == 0:
        print('[scCloseDevice] success ScStatus({}). The device index: {}'.format(str(ret), str(i)))
    else:
        print('[scCloseDevice] fail ScStatus({}). The device index: {}'.format(str(ret), str(i)))
        exit(1)

exit(0)